File: desktop_app/protocol.py
"""Versioned ASCII/CRC16 wire protocol shared by serial and demo transports."""

import binascii
from dataclasses import dataclass
import re

from .model import ArraySpec, Config, MODES, STATES

MAX_LINE = 4096
VERSION = 2
MAGIC = "HAP2"
TOKEN = re.compile(r"^[A-Za-z0-9_,.?:+\-]+$")


@dataclass(frozen=True)
class Frame:
    kind: str
    seq: int
    verb: str
    fields: dict


def encode(kind, seq, verb, **fields):
    if kind not in ("CMD", "ACK", "ERR", "TEL") or not 0 <= seq <= 65535:
        raise ValueError("帧类型或序号错误")
    tokens = [MAGIC, kind, str(seq), verb]
    for key, value in fields.items():
        if not re.fullmatch(r"[a-z][a-z0-9_]*", key) or not TOKEN.fullmatch(str(value)):
            raise ValueError("协议字段包含非法字符")
        tokens.append(f"{key}={value}")
    if not TOKEN.fullmatch(verb):
        raise ValueError("命令非法")
    body = " ".join(tokens).encode("ascii")
    result = body + f"*{binascii.crc_hqx(body, 0xFFFF):04X}\n".encode("ascii")
    if len(result) > MAX_LINE:
        raise ValueError("报文过长")
    return result


def decode(raw):
    if len(raw) > MAX_LINE:
        raise ValueError("报文过长")
    try:
        line = raw.rstrip(b"\r\n")
        body, checksum = line.rsplit(b"*", 1)
        if len(checksum) != 4 or not re.fullmatch(b"[0-9A-Fa-f]{4}", checksum):
            raise ValueError("CRC 格式错误")
        if binascii.crc_hqx(body, 0xFFFF) != int(checksum, 16):
            raise ValueError("CRC 校验失败")
        tokens = body.decode("ascii").split(" ")
        if len(tokens) < 4 or tokens[0] != MAGIC:
            raise ValueError("协议版本错误")
        kind, seq, verb = tokens[1], int(tokens[2]), tokens[3]
        if kind not in ("CMD", "ACK", "ERR", "TEL") or not 0 <= seq <= 65535:
            raise ValueError("帧头错误")
        if not TOKEN.fullmatch(verb):
            raise ValueError("命令错误")
        fields = {}
        for token in tokens[4:]:
            key, value = token.split("=", 1)
            if key in fields or not re.fullmatch(r"[a-z][a-z0-9_]*", key) or not TOKEN.fullmatch(value):
                raise ValueError("重复或非法字段")
            fields[key] = value
        return Frame(kind, seq, verb, fields)
    except (UnicodeError, IndexError, TypeError) as error:
        raise ValueError("无效报文") from error


class Decoder:
    """Bounded line reassembly; discard a whole oversized line, not just its prefix."""

    def __init__(self):
        self.buffer = bytearray()
        self.dropping = False

    def feed(self, data):
        frames, errors = [], []
        for byte in data:
            if self.dropping:
                if byte == 10:
                    self.dropping = False
                continue
            self.buffer.append(byte)
            if len(self.buffer) > MAX_LINE:
                self.buffer.clear()
                self.dropping = byte != 10
                errors.append("超长报文已丢弃")
            elif byte == 10:
                raw = bytes(self.buffer)
                self.buffer.clear()
                try:
                    frames.append((decode(raw), raw))
                except ValueError as error:
                    errors.append(str(error))
        return frames, errors


@dataclass(frozen=True)
class Snapshot:
    boot: str
    sample: int
    uptime_ms: int
    revision: int
    mode: str
    state: str
    output: bool
    config: Config
    array: ArraySpec
    focus_mm: tuple
    phases: tuple
    simulated: bool
    reason: str

    @classmethod
    def parse(cls, fields):
        config = Config.from_wire(fields)
        array = ArraySpec.from_wire(fields)
        if fields["mode"] not in MODES or fields["state"] not in STATES:
            raise ValueError("设备状态非法")
        if fields["output"] not in ("0", "1") or fields["simulated"] not in ("0", "1"):
            raise ValueError("输出标志非法")
        output = fields["output"] == "1"
        if output and (fields["state"] != "RUNNING" or config.level == 0):
            raise ValueError("输出使能与状态矛盾")
        focus = tuple(int(fields[key]) / 1000 for key in ("fx_um", "fy_um", "fz_um"))
        if not (-400 <= focus[0] <= 400 and -400 <= focus[1] <= 400 and 20 <= focus[2] <= 300):
            raise ValueError("焦点回读越界")
        phases = tuple(int(value) for value in fields["phases"].split(","))
        if len(phases) != array.count or any(not 0 <= x < config.phase_steps for x in phases):
            raise ValueError("相位快照长度或数值错误")
        counters = [int(fields[k]) for k in ("sample", "uptime_ms", "rev")]
        if min(counters) < 0 or not fields["boot"]:
            raise ValueError("状态计数非法")
        return cls(fields["boot"], *counters, fields["mode"], fields["state"], output,
                   config, array, focus, phases, fields["simulated"] == "1", fields.get("reason", "NONE"))

File: mobile_app/lib/protocol.dart
import 'dart:convert';
import 'dart:math';
import 'model.dart';

const maxFrameBytes = 8192;
final _token = RegExp(r'^[A-Za-z0-9_,.?:+|\-]+$');
final _key = RegExp(r'^[a-z][a-z0-9_]*$');

int crc16(List<int> bytes) {
  var crc = 0xffff;
  for (final byte in bytes) {
    crc ^= byte << 8;
    for (var i = 0; i < 8; i++) {
      crc = ((crc & 0x8000) != 0 ? (crc << 1) ^ 0x1021 : crc << 1) & 0xffff;
    }
  }
  return crc;
}

class Frame {
  final String kind, verb;
  final int seq;
  final Map<String, String> fields;
  const Frame(this.kind, this.seq, this.verb, [this.fields = const {}]);
  List<int> encode() {
    if (!['CMD', 'ACK', 'ERR', 'TEL'].contains(kind) ||
        seq < 0 ||
        seq > 65535 ||
        !_token.hasMatch(verb)) {
      throw const FormatException('无效帧头');
    }
    final parts = ['HAP3', kind, '$seq', verb];
    for (final e in fields.entries) {
      if (!_key.hasMatch(e.key) || !_token.hasMatch(e.value)) {
        throw const FormatException('非法协议字段');
      }
      parts.add('${e.key}=${e.value}');
    }
    final body = ascii.encode(parts.join(' '));
    final bytes = [
      ...body,
      ...ascii.encode(
        '*${crc16(body).toRadixString(16).padLeft(4, '0').toUpperCase()}\n',
      ),
    ];
    if (bytes.length > maxFrameBytes) throw const FormatException('报文过长');
    return bytes;
  }

  factory Frame.decode(List<int> bytes) {
    if (bytes.length > maxFrameBytes) throw const FormatException('报文过长');
    final line = ascii.decode(bytes).replaceFirst(RegExp(r'[\r\n]+$'), '');
    final cut = line.lastIndexOf('*');
    if (cut < 0 ||
        !RegExp(r'^[0-9a-fA-F]{4}$').hasMatch(line.substring(cut + 1))) {
      throw const FormatException('缺少校验');
    }
    if (crc16(ascii.encode(line.substring(0, cut))) !=
        int.parse(line.substring(cut + 1), radix: 16)) {
      throw const FormatException('校验失败');
    }
    final parts = line.substring(0, cut).split(' ');
    if (parts.length < 4 || parts[0] != 'HAP3') {
      throw const FormatException('协议版本不匹配');
    }
    final fields = <String, String>{};
    for (final item in parts.skip(4)) {
      final pair = item.split('=');
      if (pair.length != 2 ||
          fields.containsKey(pair[0]) ||
          !_key.hasMatch(pair[0]) ||
          !_token.hasMatch(pair[1])) {
        throw const FormatException('重复或非法字段');
      }
      fields[pair[0]] = pair[1];
    }
    final frame = Frame(
      parts[1],
      integer(parts[2]),
      parts[3],
      Map.unmodifiable(fields),
    );
    frame.encode();
    return frame;
  }
}

class FrameDecoder {
  final List<int> _buffer = [];
  bool _dropping = false;
  int rejected = 0;
  List<Frame> feed(List<int> bytes) {
    final frames = <Frame>[];
    for (final byte in bytes) {
      if (_dropping) {
        if (byte == 10) _dropping = false;
        continue;
      }
      _buffer.add(byte);
      if (_buffer.length > maxFrameBytes) {
        _buffer.clear();
        _dropping = byte != 10;
        rejected++;
      } else if (byte == 10) {
        try {
          frames.add(Frame.decode(_buffer));
        } on FormatException {
          rejected++;
        }
        _buffer.clear();
      }
    }
    return frames;
  }
}

class Snapshot {
  final FigureConfig config;
  final ArraySpec array;
  final String boot, mode, state, reason;
  final int sample, uptimeMs, revision, stroke;
  final bool output, simulated, scanOn;
  final Point<double> focus;
  final int zUm;
  final List<int> phases;
  Snapshot(Map<String, String> f)
    : config = FigureConfig.parse(f),
      array = ArraySpec.parse(f),
      boot = f['boot'] ?? '',
      mode = f['mode'] ?? '',
      state = f['state'] ?? '',
      reason = f['reason'] ?? 'NONE',
      sample = integer(f['sample']),
      uptimeMs = integer(f['uptime_ms']),
      revision = integer(f['rev']),
      stroke = integer(f['stroke_index']),
      output = f['output'] == '1',
      simulated = f['simulated'] == '1',
      scanOn = f['scan_on'] == '1',
      focus = Point(integer(f['fx_um']) / 1000, integer(f['fy_um']) / 1000),
      zUm = integer(f['fz_um']),
      phases = (f['phases'] ?? '').split(',').map(integer).toList() {
    if (boot.isEmpty ||
        min(sample, min(uptimeMs, revision)) < 0 ||
        !['LOCAL', 'REMOTE'].contains(mode) ||
        !['IDLE', 'RUNNING', 'PAUSED', 'FAULT'].contains(state) ||
        [
          'output',
          'simulated',
          'scan_on',
        ].any((key) => !['0', '1'].contains(f[key]))) {
      throw const FormatException('设备状态无效');
    }
    final strokes = config.shape == 'CUSTOM' && config.strokes.isNotEmpty
        ? config.strokes.length
        : 1;
    if (stroke < 0 ||
        stroke >= strokes ||
        (output &&
            (state != 'RUNNING' || config.value('level') == 0 || !scanOn)) ||
        focus.x.abs() > 400 ||
        focus.y.abs() > 400 ||
        zUm < 20000 ||
        zUm > 300000 ||
        phases.length != array.count ||
        phases.any((n) => n < 0 || n >= config.value('phase_steps'))) {
      throw const FormatException('设备回传不完整或相互矛盾');
    }
  }
}
